rfcs/XXXX-gene-sets/text.md

```diff
@@ -0,0 +1,100 @@
+data format for storing references to these genes and any accompanying statistical data. 
+
+## [Optional] Product Requirements
+### Users should be able to export differential expression results
+Currently cellxgene users can choose two sets of cells and compute/display the log fold and adjusted p-value of the top ten differentially expressed genes. Additionally we display a histogram of the gene expression levels of the selected cells for each gene in the right side bar. 
+- [US1] A user should be able to export the list of 10 genes along with their logfold and adjusted p-values in a csv file (see data formats below for more detail) by clicking a button
+### Users should be able to upload their own gene sets and see them in the right side bar with the option to apply them within current cellxgene functionalities, if the same gene ontology was used when generating the current dataset and the geneset uploaded by the user (ie the gene names match) 
+- [US2] A user should be able to upload a geneset via a csv file (see data formats below for more detail) for a dataset.
+- [US2.5] A user should be able to upload a geneset they previously downloaded from cellxgene
+- [UC3] When a user uploads a geneset, that list of genes should appear in the right side bar 
+- [US4] If the user was logged in when they uploaded a geneset that genesit should persist if they logout and back in or reload the page
+- [US5] If the gene names match those used in the dataset the user should be able to use general cellxgene functionality (color by genes, calculate differential expression levels, etc.) on those genes
+- [US6] Genesets uploaded for one dataset should be available in other datasets?
+
+### If there is a standard format for differential expression output, we should follow that. If not, the format should be as straightforward as possible while serving these above cases.
+
+### [Optional] Nonfunctional requirements
+- Users should be able to export precomputed differential expression results in real time
+- Users should be able to upload gene sets and have them immediately appear in the right side bar
+
+## Detailed Design | Architecture | Implementation
+
+
+### [Optional] Data model
+
+Fully supporting persistence of differential expression results is out of the scope of this RFC. But the data model described below has been designed to allow for differential expression persistence in the future.
+###Relational Database
+To support gene sets five tables will be added to the cellxgene relational database
+- Gene
+   - UUID (generated)
+   - GeneName
+   - OntologyId?
+- GeneGeneSetLink
+   - UUID
+   - Gene UUID
+   - GeneSet UUID
+   - Comments (optional)
+- GeneSet
+   - UUID (generated)
+   - GeneSet Name (required)
+   - Gene Count (generated)
+- GeneSetUserLink
+   - User UUID
+   - GeneSet UUID
+   - Comments (optional)
+   - Dataset UUID?
+-GeneSetDatasetLink
```

Comment:
Do we need a separate table for the GeneSet and Dataset relationship? One geneset belongs to only one dataset since it might have labels attached to it that refer to the dataset labels. In this case, you could have a foreign key to dataset_uuid in the geneset table itself.

Reply:
@signechambers1 mentioned wanting the genesets a user uploads to be available in multiple datasets but that requirement may have changed. If a user uploads a csv with just the geneset name and the genes (no diff exp values or category.labels) we could allow that geneset to be used in multiple datasets.

Reply:
or we could just create a copy of the geneset and link it to a different dataset

Reply:
yeah the second option of creating a copy of the geneset and linking it to a different dataset makes sense to me, but defer to @signechambers1 for guidance.

Reply:
I think creating a copy of the gene set and linking it to a different dataset makes sense to me. this seems like it cleanly supports a gene set that a user copies over from another data set and a gene set created from differential expression.

Relatedly, we probably want to include the dataset name when we support export.

Reply:
if every geneset is linked to a dataset how should we mark genesets that are uploaded with a dataset that we want to always include when displaying that dataset? We could not specify a user on those genesets but that seems fairly nonintuitive.